File: mobile/android/app/src/main/java/com/volttracker/obdpoc/data/ObdSessionStore.java
package com.volttracker.obdpoc.data;

import com.volttracker.obdpoc.materialize.ChargeSession;
import com.volttracker.obdpoc.materialize.Trip;
import java.io.Closeable;
import java.util.List;
import org.json.JSONObject;

/**
 * Write- and lifecycle-side of the on-device SQLite store. New callers that only need to write (or
 * only need to drive the session lifecycle) should depend on this narrower interface rather than
 * the full {@link ObdLocalStore} facade — it shrinks the surface they have to understand and makes
 * JVM-side fakes practical to write.
 *
 * <p>Both this and {@link ObdQueryStore} are implemented by {@link ObdLocalStore}; existing callers
 * can keep depending on the concrete facade. The split is additive; nothing was moved out of {@link
 * ObdLocalStore}.
 *
 * <p>Maintenance methods ({@link #clearAllData}, {@link #checkpoint}, {@link
 * #pruneRawDataOlderThan}) belong on the write side because they mutate state.
 */
public interface ObdSessionStore extends Closeable {

    // ---- session lifecycle ---------------------------------------------------------

    long startSession(String mode, String adapterAddress, String adapterName);

    long startSession(String mode, String adapterAddress, String adapterName, long startedAtMs);

    void finishSession(long sessionId, String status);

    void finishSession(long sessionId, String status, long endedAtMs, String supportedPids);

    void finalizeSession(
            long sessionId,
            String status,
            long endedAtMs,
            String supportedPids,
            String address,
            String adapterName,
            String mode,
            int sampleCount,
            String lastEventDetail);

    // ---- per-sample writes ---------------------------------------------------------

    long recordTelemetry(long sessionId, JSONObject sample);

    long recordTelemetry(long sessionId, JSONObject sample, long capturedAtMs);

    long recordPidObservation(long sessionId, JSONObject observation);

    long recordPidObservation(long sessionId, JSONObject observation, long observedAtMs);

    long recordPidObservation(
            long sessionId,
            long observedAtMs,
            String command,
            String header,
            String pid,
            String name,
            String valueText,
            Double valueNumeric,
            String unit,
            String rawRequest,
            String rawResponse);

    long recordDiagnosticCode(long sessionId, JSONObject diagnosticCode);

    long recordLocationSample(long sessionId, JSONObject sample);

    long recordLocationSample(long sessionId, JSONObject sample, long capturedAtMs);

    long recordLocationSample(
            long sessionId,
            long capturedAtMs,
            String provider,
            double latitude,
            double longitude,
            Double accuracyM,
            Double altitudeM,
            Double speedMps,
            Double bearingDeg,
            Long locationAgeMs,
            Long elapsedRealtimeNanos);

    long recordStatus(
            long sessionId, String state, String detail, boolean blocked, JSONObject payload);

    long recordEvent(
            long sessionId,
            String kind,
            String state,
            String detail,
            boolean blocked,
            JSONObject payload);

    long upsertVehicleFromVin(String vin);

    void recordAdapterSummary(
            String address,
            String name,
            String mode,
            long sessionId,
            String status,
            int samples,
            String supportedPids,
            String lastEventDetail);

    // ---- materializer outputs ------------------------------------------------------

    void persistTrips(long sessionId, List<Trip> trips);

    void persistChargeSessions(long sessionId, List<ChargeSession> sessions);

    // ---- maintenance ---------------------------------------------------------------

    void clearAllData();

    void checkpoint();

    int pruneRawDataOlderThan(int keepDays);

    @Override
    void close();
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/data/ObdSessionStore.kt
package com.volttracker.obdpoc.data

import com.volttracker.obdpoc.materialize.ChargeSession
import com.volttracker.obdpoc.materialize.Trip
import org.json.JSONObject
import java.io.Closeable

/** Write- and lifecycle-side of the on-device SQLite store. */
interface ObdSessionStore : Closeable {
    fun startSession(
        mode: String?,
        adapterAddress: String?,
        adapterName: String?,
    ): Long

    fun startSession(
        mode: String?,
        adapterAddress: String?,
        adapterName: String?,
        startedAtMs: Long,
    ): Long

    fun finishSession(
        sessionId: Long,
        status: String?,
    )

    fun finishSession(
        sessionId: Long,
        status: String?,
        endedAtMs: Long,
        supportedPids: String?,
    )

    fun finalizeSession(
        sessionId: Long,
        status: String?,
        endedAtMs: Long,
        supportedPids: String?,
        address: String?,
        adapterName: String?,
        mode: String?,
        sampleCount: Int,
        lastEventDetail: String?,
    )

    fun recordTelemetry(
        sessionId: Long,
        sample: JSONObject?,
    ): Long

    fun recordTelemetry(
        sessionId: Long,
        sample: JSONObject?,
        capturedAtMs: Long,
    ): Long

    fun recordPidObservation(
        sessionId: Long,
        observation: JSONObject?,
    ): Long

    fun recordPidObservation(
        sessionId: Long,
        observation: JSONObject?,
        observedAtMs: Long,
    ): Long

    fun recordPidObservation(
        sessionId: Long,
        observedAtMs: Long,
        command: String?,
        header: String?,
        pid: String?,
        name: String?,
        valueText: String?,
        valueNumeric: Double?,
        unit: String?,
        rawRequest: String?,
        rawResponse: String?,
    ): Long

    fun recordDiagnosticCode(
        sessionId: Long,
        diagnosticCode: JSONObject?,
    ): Long

    fun recordLocationSample(
        sessionId: Long,
        sample: JSONObject?,
    ): Long

    fun recordLocationSample(
        sessionId: Long,
        sample: JSONObject?,
        capturedAtMs: Long,
    ): Long

    fun recordLocationSample(
        sessionId: Long,
        capturedAtMs: Long,
        provider: String?,
        latitude: Double,
        longitude: Double,
        accuracyM: Double?,
        altitudeM: Double?,
        speedMps: Double?,
        bearingDeg: Double?,
        locationAgeMs: Long?,
        elapsedRealtimeNanos: Long?,
    ): Long

    fun recordStatus(
        sessionId: Long,
        state: String?,
        detail: String?,
        blocked: Boolean,
        payload: JSONObject?,
    ): Long

    fun recordEvent(
        sessionId: Long,
        kind: String?,
        state: String?,
        detail: String?,
        blocked: Boolean,
        payload: JSONObject?,
    ): Long

    fun upsertVehicleFromVin(vin: String?): Long

    fun recordAdapterSummary(
        address: String?,
        name: String?,
        mode: String?,
        sessionId: Long,
        status: String?,
        samples: Int,
        supportedPids: String?,
        lastEventDetail: String?,
    )

    fun persistTrips(
        sessionId: Long,
        trips: List<Trip>?,
    )

    fun persistChargeSessions(
        sessionId: Long,
        sessions: List<ChargeSession>?,
    )

    fun clearAllData()

    fun checkpoint()

    fun pruneRawDataOlderThan(keepDays: Int): Int

    override fun close()
}
